spark/spark-version/2.0/src/main/scala/org/apache/spark/storage/BlockManagerWrapper.scala

```diff
@@ -57,4 +60,48 @@ object BlockManagerWrapper {
       blockInfoManager.unlock(blockId)
     }
   }
+
+  private val putBytesMethod = {
+    val bmClass = classOf[BlockManager]
+    // Spark 2.0.0 - 2.1.0, and 2.2.0+ (as of this writing), declare the method:
+    // def putBytes[T: ClassTag](
+    //   blockId: BlockId,
+    //   bytes: ChunkedByteBuffer,
+    //   level: StorageLevel,
+    //   tellMaster: Boolean = true): Boolean
+    try {
+      bmClass.getMethod("putBytes",
+        classOf[BlockId], classOf[ChunkedByteBuffer], classOf[StorageLevel],
+        classOf[Boolean], classOf[ClassTag[_]])
+    } catch {
+      case _: NoSuchMethodException =>
+        // But Spark 2.1.1 and distros like Cloudera 2.0.0 / 2.1.0 had an extra boolean
+        // param:
+        //   def putBytes[T: ClassTag](
+        //     blockId: BlockId,
+        //     bytes: ChunkedByteBuffer,
+        //     level: StorageLevel,
+        //     tellMaster: Boolean = true,
+        //     encrypt: Boolean = false): Boolean
+        bmClass.getMethod("putBytes",
+          classOf[BlockId], classOf[ChunkedByteBuffer], classOf[StorageLevel],
+          classOf[Boolean], classOf[Boolean], classOf[ClassTag[_]])
+    }
+  }
+  private val putBytesArgCount = putBytesMethod.getParameterTypes.length
+
+  private def doPutBytes(
+      blockId: BlockId,
+      bytes: ChunkedByteBuffer,
+      level: StorageLevel): Unit = {
+    putBytesArgCount match {
```

Comment:
can we use a function object here instead of using match in every method call?

Something like this
```scala
  private val putBytes: (BlockId, ChunkedByteBuffer, StorageLevel) => Unit
      = putBytesMethod.getParameterTypes.length match {
    case 5 =>
      (blockId: BlockId, bytes: ChunkedByteBuffer, level: StorageLevel) =>
        putBytesMethod.invoke(SparkEnv.get.blockManager,
          blockId, bytes, level, JBoolean.TRUE, null)
    case 6 =>
      (blockId: BlockId, bytes: ChunkedByteBuffer, level: StorageLevel) =>
        putBytesMethod.invoke(SparkEnv.get.blockManager,
          blockId, bytes, level, JBoolean.TRUE, JBoolean.FALSE, null)
    case _ =>
      throw new IllegalArgumentException(s"invalid putBytesArgCount " +
        s"${putBytesMethod.getParameterTypes.length}")
  }

  private def doPutBytes(
      blockId: BlockId,
      bytes: ChunkedByteBuffer,
      level: StorageLevel): Unit = {
    putBytes(blockId, bytes, level)
  }
```

I write a trival program to test the performance difference, match spend a little more time than function object:
Sum 2000000, Match total time cost is 5.243864ms
Sum 2000000, Funtion object total time cost is 4.684357ms